### PR TITLE
Feature/103 improve log reporting

### PR DESCRIPTION
**Pull request by joseferben**

### CODE REVIEW

**Comment by joseferben** [comment]
@aronerben Could you have a look at this PR? It's the first iteration to improve log reporting so it's a bit easier to 1) find issues during development and 2) find issues during production so that bugs can be fixed just by looking at logs.

https://gitlab.uzh.ch/econ/study-coordination/pool/-/issues/103

Once we convert `Lwt.Syntax` to `lwt_ppx` we can attach a useful backtrace every time `Logs.err` is called, that should be very useful too.

**Comment by joseferben** [comment]
We need to make sure that it's not possible to accidentally print secrets, especially around custom fields and so on. I think if we encrypt logs we should be able to print sensitive personal stuff.

File: pool/run/run.ml [hunk 1]
```diff
@@ -39,5 +42,5 @@ let () =
     |> before_start (fun () ->
          let () = Middleware.Error.before_start () in
          Printexc.record_backtrace true |> Lwt.return)
```
Comment (joseferben) [inline pool/run/run.ml:44]: Remove Printexec here

**Review by aronerben — CHANGES_REQUESTED** [review]

**Comment by aronerben** [comment]
These are general comments, will give a more in-depth look soon

File: pool/app/logger/logger.ml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+let pp_source = Fmt.(styled source_style string)
+let pp_context = Fmt.(styled `Green string)
+let pp_id = Fmt.(styled `Red string)
+
+let pp_exec_header tags src =
+  let pp_h ppf style h =
+    let id =
+      tags
+      |> Option.map @@ Logs.Tag.find tag_req
```
[the remark this answers is not in the document]
Comment (joseferben) [reply pool/app/logger/logger.ml:65]: Actually a helpful heuristic, thanks!

**Comment by joseferben** [comment]
@aronerben I fixed those general topics.

File: pool/app/pool_common/pool_common_utils.ml [hunk 3]
```diff
@@ -13,40 +19,88 @@ module Error = struct
...
 end
 
-module PoolConformist = Conformist.Make (Error)
+module PoolConformist = struct
+  include Conformist.Make (Error)
+
+  let pp_schema (f : Format.formatter) (schema, input) =
+    let schema =
+      fold_left
+        ~f:(fun res field -> List.cons (Field.name field) res)
```
[the remark this answers is not in the document]
Comment (aronerben) [reply pool/app/pool_common/pool_common_utils.ml:28]: String, List, Format, Result, Option

File: pool/cqrs_command/pool_tenant_command.mli [hunk 4]
```diff
@@ -0,0 +1,107 @@
+module Conformist = Pool_common.Utils.PoolConformist
```
Comment (aronerben) [inline pool/cqrs_command/pool_tenant_command.mli:1]: Why does this file exist? The interfaces are inlined in the `ml` file

File: pool/run/run.ml (hunk 1 above)
Comment (aronerben) [reply pool/run/run.ml:44]: Is this done?

File: pool/test/tenant_settings_test.ml [hunk 5]
```diff
@@ -145,7 +145,10 @@ let login_after_terms_update _ () =
       let%lwt accepted = Contact.has_terms_accepted database_label contact in
       match accepted with
       | true -> Lwt.return_ok contact
-      | false -> Lwt.return_error TermsAndConditionsNotAccepted
+      | false ->
+        Logs.warn (fun m ->
```
Comment (aronerben) [inline pool/test/tenant_settings_test.ml:149]: I don't think we need this in tests

File: pool/web/handler/admin_experiments_waiting_list.ml [hunk 6]
```diff
@@ -172,14 +175,17 @@ let assign_contact req =
       let session_text = Session.(to_email_text language session) in
       Lwt_result.return Email.{ subject; text; language; session_text }
     in
+    let tags = Logger.req req in
     let events =
       Cqrs_command.Assignment_command.CreateFromWaitingList.(
-        handle { session; waiting_list; already_enrolled })
+        handle ~tags { session; waiting_list; already_enrolled })
```
Comment (aronerben) [inline pool/web/handler/admin_experiments_waiting_list.ml:181]: Did you check that `tags` are passed in for every single command handler?

File: pool/web/middleware/middleware_logger.ml [hunk 7]
```diff
@@ -0,0 +1,119 @@
+(* Based on
```
Comment (aronerben) [inline pool/web/middleware/middleware_logger.ml:1]: Is this an improved version? Should we PR this to them?

File: pool/web/middleware/middleware_tenant.ml [hunk 8]
```diff
@@ -18,7 +18,10 @@ let valid_tenant () =
     in
     match result with
     | Ok context -> context |> Pool_context.Tenant.set req |> handler
-    | Error _ -> Http_utils.redirect_to "/not-found"
+    | Error _ ->
+      Logs.err (fun m ->
+        m "Not pool context found in request" ~tags:(Logger.req req));
```
Comment (aronerben) [inline pool/web/middleware/middleware_tenant.ml:23]: Typo, "No"

File: pool/web/utils/http_utils.ml [hunk 9]
```diff
@@ -56,31 +56,43 @@ let redirect_to path = redirect_to_with_actions path []
...
-    |> Pool_common.Utils.with_log_result_error (fun (err, _) -> err)
+    |> Pool_common.Utils.with_log_result_error ~tags (fun (err, _) -> err)
     |> CCResult.map Lwt.return
     |> CCResult.get_lazy (fun (error_msg, error_path) ->
          redirect_to_with_actions
            (path_with_language query_language error_path)
            [ msgf error_msg ])
-  | Error _ -> redirect_to "/error"
+  | Error err ->
+    Printexc.print_backtrace stderr;
```
Comment (aronerben) [inline pool/web/utils/http_utils.ml:71]: Remove please

File: pool/web/utils/http_utils.ml [hunk 10]
```diff
@@ -56,31 +56,43 @@ let redirect_to path = redirect_to_with_actions path []
...
+    Logs.warn (fun m ->
+      m "Context not found: %s" (Message.Message.show_error err) ~tags);
+    redirect_to "/error"
 ;;
 
 let extract_happy_path req result =
   extract_happy_path_generic req result (fun err ->
+    Printexc.print_backtrace stderr;
```
Comment (aronerben) [inline pool/web/utils/http_utils.ml:79]: Remove please

File: pool/web/utils/http_utils.ml [hunk 11]
```diff
@@ -93,7 +105,11 @@ let extract_happy_path_with_actions req result =
                   ~error:[ error_key ]
               ]
               error_actions))
-  | Error _ -> redirect_to "/error"
+  | Error err ->
+    Printexc.print_backtrace stderr;
```
Comment (aronerben) [inline pool/web/utils/http_utils.ml:109]: Remove please

File: pool/web/utils/http_utils.ml [hunk 12]
```diff
@@ -121,7 +138,11 @@ let extract_happy_path_htmx req result =
            ())
   | Error err ->
     Logs.err (fun m ->
-      m "%s" @@ Pool_common.(Utils.error_to_string Language.En err));
+      m ~tags:(Logger.req req) "%s"
+      @@ Pool_common.(Utils.error_to_string Language.En err));
+    Printexc.print_backtrace stderr;
```
Comment (aronerben) [inline pool/web/utils/http_utils.ml:143]: Remove please

**Comment by aronerben** [comment]
@joseferben Please push your most recent changes

File: pool/web/handler/admin_experiments_waiting_list.ml (hunk 6 above)
Comment (joseferben) [reply pool/web/handler/admin_experiments_waiting_list.ml:177]: root_command.ml needs to be done

**Review by aronerben — APPROVED** [review]

**Pull request merged**